2801/multi-select combo box component

jaredcwhite opened the pull request:
## Issue Overview

This PR addresses #2801 

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This PR implements the control in the Figma mockup linked in the issue. The multi-select field will accept a `dataSource` prop which can be a function with returns a promise, making it ideal for lazy-loading content from the server. Or you can provide a pre-loaded dataset. (Basically any of the source options allowed by the aria-autocomplete library.)

## How Can This Be Tested/Reviewed?

You can preview the component in Storybook: https://deploy-preview-2844--dev-storybook-bloom.netlify.app/?path=/story/forms-multiselect-field--standard

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [x] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

jaredcwhite commented:
@slowbot The Storybook story is at a good point to try out…I'd still want to write up the documentation before merge.

kramduckner commented on ui-components/src/forms/MultiSelectField.tsx:
```diff
@@ -0,0 +1,67 @@
+        },
+      })
+    }
+  }, [autocompleteRef, name, setValue, props.dataSource])
```
I'm curious how we handle this in all our components - Without having a check to see if props.dataSource exists, is it possible that this useEffect will get run multiple times, for example, before and after data is fetched? Should there be a check to see it exists before attaching it to the ref? Or is that state handled in the parent component, and it's assumed you'll always have data on the first mount?

jaredcwhite replied:
I think the idea is that either you have data on the first mount, or you supply an async function so the data is loaded in real-time. We should make sure `useEffect` doesn't run more than once, unless there's a major change to those form control details. The `dataSource` prop is definitely required, otherwise I believe the aria-autocomplete control won't work correctly.

jaredcwhite commented on ui-components/src/forms/MultiSelectField.scss:
```diff
@@ -0,0 +1,220 @@
+  box-shadow: none;
+}
+
+.multi-select-field__option:hover,
```
Interesting question…arguably they're both a list of items, although in one case it's a list of links, and in another a list of menu items. Unfortunately I don't have much control over the exact class names—the aria-autocomplete library is generating a bunch of the names. But I could see us defining some higher-level design tokens and using those as variables for this and other similar lists.

jaredcwhite commented:
Sounds good @slowbot, I'll work on those changes.

jaredcwhite commented:
Sorta lost track of this over the holiday break…will try to wrap it up again.

dominikx96 reviewed (changes requested):
@jaredcwhite see comments above and:

- It would be great to have tests for this component
- a tag has incorrect close icon color on hover:
![Screenshot 2022-07-11 at 12 47 30](https://user-images.githubusercontent.com/16655908/178252259-5673f1f7-6de5-418b-8575-7e1b9daf7339.png)

jaredcwhite commented:
All right, looks like all relevant tests/builds that need to be passing are passing now. I believed I've addressed all feedback except adding a unit test for the component. I'm not sure exactly the best way to do that, since nearly all interactions are actually managed by aria-autocomplete.

emilyjablonski commented on ui-components/src/forms/MultiSelectField.tsx:
```diff
@@ -0,0 +1,92 @@
+        const AriaAutocomplete = (await import("aria-autocomplete")).default
+        AriaAutocomplete(autocompleteRef.current, {
+          source: props.dataSource,
+          delay: 500, // debounce for a half-second
```
There is no need to debounce on the first key press

emilyjablonski commented on ui-components/src/forms/MultiSelectField.scss:
```diff
@@ -0,0 +1,232 @@
+  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='10' height='10' viewBox='0 0 10 10' version='1.1'%3E%3Cg id='surface1'%3E%3Cpath d='M 0.234375 10 L 5 5.21875 L 9.765625 10 L 9.988281 9.78125 L 5.21875 5 L 9.988281 0.21875 L 9.765625 0 L 5 4.78125 L 0.234375 0 L 0.0117188 0.21875 L 4.78125 5 L 0.0117188 9.78125 ...
+}
+
+@media speech {
```
can you walk me through what the speech query does? not finding great resources online

jaredcwhite replied:
@emilyjablonski I have no idea TBH. 😅 Some of this stylesheet was pulled from the aria-autocomplete example.

emilyjablonski commented on ui-components/src/forms/MultiSelectField.tsx:
```diff
@@ -0,0 +1,92 @@
+export interface MultiSelectFieldProps {
+  name: string
+  dataSource:
+    | string
```
why not just always require a string[] ?

emilyjablonski commented on ui-components/src/forms/MultiSelectField.tsx:
```diff
@@ -0,0 +1,92 @@
+  register({ name }, props.validation)
+
+  useEffect(() => {
+    // We need to dynamically import the aria-autocomplete control only on the
```
Does this make it so this component does not work out of the box for consumers?

emilyjablonski commented:
A few questions! I also think it is important to add a simple test (does this component render, can I add an item and remove an item) to ensure it is working appropriately and that we are consuming the api appropriately.

jaredcwhite commented:
@emilyjablonski I was able to add a unit test, and also I lowered the debounce value a bit (plus actually the data source in Storybook was simulating a 500ms request delay, so I trimmed that down to 100ms). Let me know what you think!

jaredcwhite commented:
(Oh and I added a new static data source example with a large number of items to scroll.)

emilyjablonski commented on ui-components/__tests__/forms/MultiSelectField.test.tsx:
```diff
@@ -0,0 +1,61 @@
+  )
+}
+
+describe("<MultiSelectField>", () => {
```
Love this testt

emilyjablonski reviewed (commented):
This is looking so awesome!! Thank you for the changes. Everything above looks great, I just have one thing I didn't catch before (sorry about that!) which is the loading state - if I check out the Storybook preview on a throttled connection here is the loading state:

![Screen Shot 2022-07-20 at 12 28 28 PM](https://user-images.githubusercontent.com/65367387/180056198-a1f1b556-77da-478e-84de-caf69eb6505f.png)

jaredcwhite commented:
@emilyjablonski Good catch on the `undefined` issue…I was able to track that down and fix, and also make the input height stable so there's not layout shift as the control loads in. Think we're good to go now!
